### DISCUSSION TITLE
NotifyCanExecuteChangedFor不能触发绑定的命令

### DISCUSSION THREAD

**Discussion by minChaojie** [BYJRK/DotNet-Discussions#143] [Q&A]
你好，我在使用 CommunityToolkit.Mvvm 时遇到一个问题，`NotifyCanExecuteChangedFor` 属性改变后不会触发我绑定的命令，现在只能使用 `OnTestPropertyChanged` 来实现功能。

```c#
[ObservableProperty]
[NotifyCanExecuteChangedFor(nameof(BackProcessCommand))]
private string testProperty = "123";
//partial void OnTestPropertyChanged(string? oldValue, string newValue)
//{
//    TestProperty = "789";
//}
[RelayCommand]
public void BackProcess()
{
    TestProperty = "345";
}

[RelayCommand]
private void TestProcess()
{
    TestProperty = "345";
}
```

**Comment by BYJRK** [comment]
你没有为你的 `RelayCommand` 特性中的 `CanExecute` 属性赋值，应该这样写：

```c#
[ObservableProperty]
[NotifyCanExecuteChangedFor(nameof(BackProcessCommand))]
private string testProperty = "123";

bool CanBackProcessExecute() => !string.IsNullOrEmpty(TestProperty);

[RelayCommand(CanExecute = nameof(CanBackProcessExecute)]
public void BackProcess()
{
    TestProperty = "345";
}
```

**Reply by minChaojie** [reply]
嗯嗯我早上看了下RelayCommand的代码，是有个这个。谢谢了
public void NotifyCanExecuteChanged()
        {
            this.CanExecuteChanged?.Invoke(this, EventArgs.Empty);
        }

**Reply by minChaojie** [reply]
我刚才试了下还是不行，是因为我改变属性值的触发有问题？我是在一个按键命令里面去改变testProperty的值，这有影响吗？

**Reply by BYJRK** [reply]
你修改的是字段还是属性的值？

**Reply by minChaojie** [reply]
修改的属性值，就修改TestProperty的值

**Reply by minChaojie** [reply]
[ObservableProperty]
        [NotifyCanExecuteChangedFor(nameof(BackProcessCommand))]
        private string testProperty = "123";

        bool CanBackProcessExecute() => !string.IsNullOrEmpty(TestProperty);

        [RelayCommand(CanExecute = nameof(CanBackProcessExecute))]
        public void BackProcess()
        {
            MessageBox.Show("789");
        }

        [RelayCommand]
        private void TestProcess()
        {
            TestProperty = "345";
        }
现在代码是这样
